lib/Doctrine/DBAL/Driver/Connection.php

```diff
@@ -15,62 +15,72 @@
 {
     /**
      * Prepares a statement for execution and returns a Statement object.
+     *
+     * @param string $sql The SQL query to prepare.
+     *
+     * @return Statement The prepared statement.
      */
     public function prepare(string $sql) : Statement;
 
     /**
      * Executes an SQL statement, returning a result set as a Statement object.
      *
+     * @param string $sql The SQL query to execute.
+     *
+     * @return ResultStatement The result statement.
+     *
      * @throws DBALException
      */
     public function query(string $sql) : ResultStatement;
 
     /**
      * Quotes a string for use in a query.
      *
-     * @param mixed $input
-     * @param int   $type
+     * If the driver does not support quoting, an exception is thrown.
```

Comment:
We don't have any. I added this sentence because `PDO::quote()` [can return false](http://php.net/manual/en/pdo.quote.php) *"if the driver does not support quoting in this way"*. 
I have no idea what this means. Please check #3334 where I was discussing `Connection::quote()`, and was suggesting that we accept `string` only, I guess this would solve the problem as every driver should indeed be able to quote strings, and I cannot see a value in quoting other types. Then we would have no reason to say that it can throw an exception.

Reply:
The confusion comes from the method name. I think in the broadest sense, the purpose of `quote()` is to take a value and convert it into a SQL expression representing that value. Hence, the return type of `string` makes the most sense. Given the purpose, any value is acceptable (e.g. `(int) 0 => '0'`, `null => 'NULL'`). If `quote()` doesn't do that, then the client side code would have to handle everything else than strings which is tedious.

Reply:
Regarding the "if the driver does not support quoting", I'd remove it.

Reply:
I just tested `PDO::quote()`'s implementation on both sqlite and mysql:

```php
echo $pdo->quote(null); // ''
echo $pdo->quote(null, PDO::PARAM_NULL); // ''
```

They all return an empty single-quoted string. For other types (int, float), this is just basically adding single quotes around the value type-casted to string.

Basically, what the PDO implementation seems to be doing is type-casting the input to a string, then applying driver-specific quoting to that string (actually, `PDO::quote()` [explicitly says](http://php.net/manual/en/pdo.quote.php) that the input type is `string`). This is also what DBAL implementations of `Driver\Connection::quote()` currently do.

So even though what you said above makes sense, it is in all likelihood not how PDO works, and it is not how the current DBAL implementation of `quote()` works. So it looks like nobody is currently relying on this behaviour.

As such, I would still strongly suggest that the API becomes:

```php
public function quote(string $string) : string;
```

This would be much simpler and remove all confusion around this topic.

Reply:
Indeed. Out of the supported PDO drivers, only [`mysql`](https://github.com/php/php-src/blob/0f7f1498be549f5988bf9d8150f35bedf70741c5/ext/pdo_pgsql/pgsql_driver.c#L334) and [`pgsql`](https://github.com/php/php-src/blob/8d3f8ca12a0b00f2a74a27424790222536235502/ext/pdo_mysql/mysql_driver.c#L306-L311) implement _some_ type-specific logic but it's now what we support in DBAL.

Let's update the API to the suggested about.

Reply:
Cool, done in e6dcadb. I've inline-commented on 2 gotchas, though.

Reply:
2 other gotchas, that I can't inline-comment on: [static analysis](https://travis-ci.org/doctrine/dbal/jobs/456856374) showed 2 `quote()` calls with the second parameter in use:

- [ExpressionBuilder::literal()](https://github.com/doctrine/dbal/blob/e6dcadb6f4d099d297527d81617d132e28f1713a/lib/Doctrine/DBAL/Query/Expression/ExpressionBuilder.php#L296)
- [SQLAzureShardManager::splitFederation()](https://github.com/doctrine/dbal/blob/e6dcadb6f4d099d297527d81617d132e28f1713a/lib/Doctrine/DBAL/Sharding/SQLAzure/SQLAzureShardManager.php#L211)

These methods are not unit-tested, and not used anywhere in the code.

Because of the current `quote()` behaviour discussed above, I think we can safely rewrite them to not use the second parameter. What do you think?

Also, it would be nice to check if these functions are actually useful, or could be removed from the codebase altogether.

Reply:
> I think we can safely rewrite them to not use the second parameter. What do you think?

I agree. Especially in the second case, the distribution type is always a string.

> Also, it would be nice to check if these functions are actually useful, or could be removed from the codebase altogether.

Let's leave them for now. I haven't used `ExpressionBuilder::literal()` myself but it seems to make sense. Just remove the second argument from the call.

Reply:
Done! Notes:

- I changed the parameter type to `string` in both functions, as any other type does not make sense anymore
- I removed the `$type` parameter from `ExpressionBuilder::literal()` as it's unused now

Reply:
:+1: